File: apps/api/src/teams/teams.service.ts
import { randomBytes } from 'node:crypto';
import {
  BadRequestException,
  ConflictException,
  ForbiddenException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Prisma, UserRole } from '@prisma/client';
import type {
  AssignMemberInput,
  AssignMemberResult,
  AssignableUser,
  CreateTeamInput,
  TeamDetail,
  TeamMember,
  TeamRole,
  TeamSummary,
  UpdateTeamInput,
} from '@ihrms/shared';
import type { Env } from '../config/env.validation';
import { PrismaService } from '../prisma/prisma.service';
import { AuditService } from '../audit/audit.service';
import { MailService } from '../auth/mail.service';
import { hashSecret } from '../auth/hashing';
import type { UserPrincipal } from '../auth/principal';

const MEMBER_SELECT = {
  id: true,
  name: true,
  email: true,
  role: true,
  status: true,
} satisfies Prisma.UserSelect;

interface MemberRow {
  id: string;
  name: string;
  email: string;
  role: UserRole;
  status: string;
}

interface TeamRow {
  id: string;
  name: string;
  createdAt: Date;
  hr: MemberRow | null;
  manager: MemberRow | null;
  _count: { members: number };
}

function toMember(user: MemberRow): TeamMember {
  return {
    id: user.id,
    name: user.name,
    email: user.email,
    role: user.role as TeamMember['role'],
    status: user.status,
  };
}

function toSummary(team: TeamRow): TeamSummary {
  return {
    id: team.id,
    name: team.name,
    hr: team.hr ? toMember(team.hr) : null,
    manager: team.manager ? toMember(team.manager) : null,
    memberCount: team._count.members,
    createdAt: team.createdAt.toISOString(),
  };
}

/**
 * Team management (§2/§3.1). COMPANY_ADMIN-only (controller guard); every query is
 * filtered by the admin's `companyId` so there is no cross-tenant access. The single
 * hrUserId/managerUserId columns structurally enforce exactly one HR + one Manager.
 */
@Injectable()
export class TeamsService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly audit: AuditService,
    private readonly mail: MailService,
    private readonly config: ConfigService<Env, true>,
  ) {}

  async create(input: CreateTeamInput, actor: UserPrincipal, ip?: string): Promise<TeamDetail> {
    const companyId = this.companyOf(actor);
    const team = await this.prisma.guarded.team.create({
      data: { name: input.name, companyId },
      select: { id: true },
    });
    await this.audit.record({
      actorType: 'USER',
      actorId: actor.userId,
      companyId,
      action: 'TEAM_CREATED',
      targetType: 'Team',
      targetId: team.id,
      metadata: { name: input.name },
      ipAddress: ip ?? null,
    });
    return this.getDetail(team.id, actor);
  }

  async list(actor: UserPrincipal): Promise<TeamSummary[]> {
    const companyId = this.companyOf(actor);
    const rows = await this.prisma.guarded.team.findMany({
      where: { companyId },
      orderBy: { createdAt: 'desc' },
      select: {
        id: true,
        name: true,
        createdAt: true,
        hr: { select: MEMBER_SELECT },
        manager: { select: MEMBER_SELECT },
        _count: { select: { members: true } },
      },
    });
    return rows.map(toSummary);
  }

  async getDetail(id: string, actor: UserPrincipal): Promise<TeamDetail> {
    const companyId = this.companyOf(actor);
    const team = await this.prisma.guarded.team.findFirst({
      where: { id, companyId },
      select: {
        id: true,
        name: true,
        createdAt: true,
        hr: { select: MEMBER_SELECT },
        manager: { select: MEMBER_SELECT },
        _count: { select: { members: true } },
        members: { select: MEMBER_SELECT, orderBy: { name: 'asc' } },
      },
    });
    if (!team) {
      throw new NotFoundException('Team not found');
    }
    return { ...toSummary(team), members: team.members.map(toMember) };
  }

  async update(
    id: string,
    input: UpdateTeamInput,
    actor: UserPrincipal,
    ip?: string,
  ): Promise<TeamDetail> {
    const companyId = this.companyOf(actor);
    await this.ensureTeam(id, companyId);
    await this.prisma.guarded.team.update({ where: { id }, data: { name: input.name } });
    await this.audit.record({
      actorType: 'USER',
      actorId: actor.userId,
      companyId,
      action: 'TEAM_UPDATED',
      targetType: 'Team',
      targetId: id,
      metadata: { name: input.name },
      ipAddress: ip ?? null,
    });
    return this.getDetail(id, actor);
  }

  async remove(id: string, actor: UserPrincipal, ip?: string): Promise<void> {
    const companyId = this.companyOf(actor);
    await this.ensureTeam(id, companyId);
    const approvals = await this.prisma.guarded.approvalRequest.count({ where: { teamId: id } });
    if (approvals > 0) {
      throw new ConflictException('Team has approval history and cannot be deleted');
    }
    // Detach members (clears their teamId), then delete the team.
    await this.prisma.guarded.$transaction([
      this.prisma.guarded.user.updateMany({ where: { teamId: id }, data: { teamId: null } }),
      this.prisma.guarded.team.delete({ where: { id } }),
    ]);
    await this.audit.record({
      actorType: 'USER',
      actorId: actor.userId,
      companyId,
      action: 'TEAM_DELETED',
      targetType: 'Team',
      targetId: id,
      ipAddress: ip ?? null,
    });
  }

  async assignableUsers(actor: UserPrincipal, role: string): Promise<AssignableUser[]> {
    const companyId = this.companyOf(actor);
    if (role !== UserRole.HR && role !== UserRole.MANAGER) {
      throw new BadRequestException('role must be HR or MANAGER');
    }
    const users = await this.prisma.guarded.user.findMany({
      where: { companyId, role, teamId: null },
      orderBy: { name: 'asc' },
      select: MEMBER_SELECT,
    });
    return users.map(toMember);
  }

  /** Assign the HR or Manager slot by selecting an existing user or creating a new one. */
  async assign(
    id: string,
    role: TeamRole,
    input: AssignMemberInput,
    actor: UserPrincipal,
    ip?: string,
  ): Promise<AssignMemberResult> {
    const companyId = this.companyOf(actor);
    const team = await this.prisma.guarded.team.findFirst({
      where: { id, companyId },
      select: { id: true, hrUserId: true, managerUserId: true },
    });
    if (!team) {
      throw new NotFoundException('Team not found');
    }

    const isHr = role === UserRole.HR;
    const currentUserId = isHr ? team.hrUserId : team.managerUserId;
    const otherUserId = isHr ? team.managerUserId : team.hrUserId;

    let userId: string;
    let devPassword: string | undefined;

    if ('userId' in input) {
      const user = await this.prisma.guarded.user.findFirst({
        where: { id: input.userId, companyId },
        select: { id: true, role: true, teamId: true },
      });
      if (!user) {
        throw new NotFoundException('User not found in your company');
      }
      if (user.role !== role) {
        throw new BadRequestException(`Selected user is not ${role}`);
      }
      if (user.teamId && user.teamId !== id) {
        throw new ConflictException('User is already assigned to another team');
      }
      userId = user.id;
    } else {
      const tempPassword = randomBytes(12).toString('base64url');
      try {
        const created = await this.prisma.guarded.user.create({
          data: {
            name: input.name,
            email: input.email,
            role,
            companyId,
            teamId: id,
            passwordHash: await hashSecret(tempPassword),
            status: 'ACTIVE',
          },
          select: { id: true },
        });
        userId = created.id;
      } catch (err) {
        if (err instanceof Prisma.PrismaClientKnownRequestError && err.code === 'P2002') {
          throw new ConflictException(`Email "${input.email}" is already in use`);
        }
        throw err;
      }
      await this.mail.sendStaffInvite(input.email, role, tempPassword);
      devPassword = this.isProd() ? undefined : tempPassword;
    }

    if (otherUserId && otherUserId === userId) {
      throw new ConflictException('This person already holds the other role on this team');
    }

    await this.prisma.guarded.$transaction(async (tx) => {
      // Detach a replaced holder so a team keeps exactly one person in the slot.
      if (currentUserId && currentUserId !== userId) {
        await tx.user.update({ where: { id: currentUserId }, data: { teamId: null } });
      }
      await tx.user.update({ where: { id: userId }, data: { teamId: id, role } });
      await tx.team.update({
        where: { id },
        data: isHr ? { hrUserId: userId } : { managerUserId: userId },
      });
    });

    await this.audit.record({
      actorType: 'USER',
      actorId: actor.userId,
      companyId,
      action: isHr ? 'TEAM_HR_ASSIGNED' : 'TEAM_MANAGER_ASSIGNED',
      targetType: 'Team',
      targetId: id,
      metadata: { userId, role },
      ipAddress: ip ?? null,
    });

    const detail = await this.getDetail(id, actor);
    return devPassword ? { team: detail, devPassword } : { team: detail };
  }

  private companyOf(actor: UserPrincipal): string {
    if (!actor.companyId) {
      throw new ForbiddenException('No company in scope');
    }
    return actor.companyId;
  }

  private async ensureTeam(id: string, companyId: string): Promise<void> {
    const team = await this.prisma.guarded.team.findFirst({
      where: { id, companyId },
      select: { id: true },
    });
    if (!team) {
      throw new NotFoundException('Team not found');
    }
  }

  private isProd(): boolean {
    return this.config.get('NODE_ENV', { infer: true }) === 'production';
  }
}

File: apps/api/src/main/java/com/ihrms/teams/TeamsService.java
package com.ihrms.teams;

import com.ihrms.audit.AuditActor;
import com.ihrms.audit.AuditService;
import com.ihrms.auth.AccountEmails;
import com.ihrms.auth.IhrmsPrincipal;
import com.ihrms.auth.MailService;
import com.ihrms.domain.enums.UserRole;
import com.ihrms.domain.model.Company;
import com.ihrms.domain.model.Team;
import com.ihrms.domain.model.User;
import com.ihrms.domain.repository.ApprovalRequestRepository;
import com.ihrms.domain.repository.CompanyRepository;
import com.ihrms.domain.repository.TeamRepository;
import com.ihrms.domain.repository.UserRepository;
import com.ihrms.mail.MailAddresses;
import com.ihrms.teams.dto.TeamDtos.AssignMemberRequest;
import com.ihrms.teams.dto.TeamDtos.AssignMemberResult;
import com.ihrms.teams.dto.TeamDtos.CreateTeamRequest;
import com.ihrms.teams.dto.TeamDtos.TeamDetailView;
import com.ihrms.teams.dto.TeamDtos.TeamMemberView;
import com.ihrms.teams.dto.TeamDtos.TeamSummaryView;
import com.ihrms.teams.dto.TeamDtos.UpdateTeamRequest;
import java.util.List;
import java.util.Map;
import org.springframework.core.env.Environment;
import org.springframework.core.env.Profiles;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

/**
 * Team management (ARCHITECTURE.md §2/§3.1). COMPANY_ADMIN-only; EVERY query is filtered by the
 * admin's companyId — no cross-tenant access. The single nullable hrUserId/managerUserId columns
 * structurally enforce exactly one HR + one Manager; a replaced holder is detached so the slot
 * keeps exactly one person.
 */
@Service
public class TeamsService {

  private final TeamRepository teams;
  private final UserRepository users;
  private final CompanyRepository companies;
  private final ApprovalRequestRepository approvals;
  private final AuditService audit;
  private final MailService mail;
  private final MailAddresses mailAddresses;
  private final PasswordEncoder encoder;
  private final AccountEmails accountEmails;
  private final Environment env;

  public TeamsService(
      TeamRepository teams,
      UserRepository users,
      CompanyRepository companies,
      ApprovalRequestRepository approvals,
      AuditService audit,
      MailService mail,
      MailAddresses mailAddresses,
      PasswordEncoder encoder,
      AccountEmails accountEmails,
      Environment env) {
    this.teams = teams;
    this.users = users;
    this.companies = companies;
    this.approvals = approvals;
    this.audit = audit;
    this.mail = mail;
    this.mailAddresses = mailAddresses;
    this.encoder = encoder;
    this.accountEmails = accountEmails;
    this.env = env;
  }

  public TeamDetailView create(
      String companyId, CreateTeamRequest input, IhrmsPrincipal.User actor, String ip) {
    Team team = new Team();
    team.setName(input.name().trim());
    team.setCompanyId(companyId);
    teams.save(team);
    audit(actor, companyId, "TEAM_CREATED", team.getId(), Map.of("name", team.getName()), ip);
    return detail(team);
  }

  public List<TeamSummaryView> list(String companyId) {
    return teams.findByCompanyIdOrderByCreatedAtDesc(companyId).stream().map(this::summary).toList();
  }

  public TeamDetailView getDetail(String companyId, String id) {
    return detail(requireTeam(id, companyId));
  }

  public TeamDetailView update(
      String companyId, String id, UpdateTeamRequest input, IhrmsPrincipal.User actor, String ip) {
    Team team = requireTeam(id, companyId);
    team.setName(input.name().trim());
    teams.save(team);
    audit(actor, companyId, "TEAM_UPDATED", id, Map.of("name", team.getName()), ip);
    return detail(team);
  }

  @Transactional
  public void remove(String companyId, String id, IhrmsPrincipal.User actor, String ip) {
    Team team = requireTeam(id, companyId);
    if (approvals.countByTeamId(id) > 0) {
      throw conflict("Team has approval history and cannot be deleted");
    }
    // Detach members (clears their teamId), then delete the team.
    for (User member : users.findByTeamIdOrderByNameAsc(id)) {
      member.setTeamId(null);
      users.save(member);
    }
    teams.delete(team);
    audit(actor, companyId, "TEAM_DELETED", id, Map.of(), ip);
  }

  public List<TeamMemberView> assignableUsers(String companyId, String role) {
    UserRole parsed = parseTeamRole(role);
    return users
        .findByCompanyIdAndRoleAndTeamIdIsNullOrderByNameAsc(companyId, parsed)
        .stream()
        .map(TeamsService::memberView)
        .toList();
  }

  /** Assign the HR or Manager slot by selecting an existing user or creating a new one. */
  @Transactional
  public AssignMemberResult assign(
      String companyId,
      String id,
      UserRole role,
      AssignMemberRequest input,
      IhrmsPrincipal.User actor,
      String ip) {
    Team team = requireTeam(id, companyId);

    String currentUserId = slotUserId(team, role);
    List<String> otherSlotUserIds = otherSlotUserIds(team, role);

    String userId;
    String devPassword = null;

    if (input.userId() != null && !input.userId().isBlank()) {
      User user =
          users
              .findByIdAndCompanyId(input.userId(), companyId)
              .orElseThrow(() -> notFound("User not found in your company"));
      if (user.getRole() != role) {
        throw badRequest("Selected user is not " + role);
      }
      if (user.getTeamId() != null && !user.getTeamId().equals(id)) {
        throw conflict("User is already assigned to another team");
      }
      userId = user.getId();
    } else {
      if (isBlank(input.name())) {
        throw badRequest("Provide a userId, or a name and mailbox name");
      }
      // The mailbox address = localPart@companyDomain, and IS the login email (§8).
      Company company =
          companies.findById(companyId).orElseThrow(() -> notFound("Company not found"));
      var address =
          mailAddresses.resolve(input.localPart(), input.email(), company.getMailDomain());
      accountEmails.assertAvailableForStaff(address.email()); // unique across staff + employees (§6)
      // New staff sign in with email + password (§6): an initial password (min 8) is required.
      String password = input.password();
      if (isBlank(password) || password.trim().length() < 8) {
        throw badRequest("An initial password of at least 8 characters is required");
      }
      User created = new User();
      created.setName(input.name().trim());
      created.setEmail(address.email());
      created.setMailLocalPart(address.localPart());
      created.setRole(role);
      created.setCompanyId(companyId);
      created.setTeamId(id);
      created.setPasswordHash(encoder.encode(password));
      created.setStatus("ACTIVE");
      try {
        users.saveAndFlush(created);
      } catch (DataIntegrityViolationException e) {
        throw conflict("Email \"" + address.email() + "\" is already in use");
      }
      userId = created.getId();
      mail.sendStaffInvite(address.email(), role.name(), password);
      devPassword = isProd() ? null : password;
    }

    if (otherSlotUserIds.contains(userId)) {
      throw conflict("This person already holds another role on this team");
    }

    // Detach a replaced holder so the slot keeps exactly one person.
    if (currentUserId != null && !currentUserId.equals(userId)) {
      users
          .findById(currentUserId)
          .ifPresent(
              prev -> {
                prev.setTeamId(null);
                users.save(prev);
              });
    }
    User assignee = users.findById(userId).orElseThrow(() -> notFound("User not found"));
    assignee.setTeamId(id);
    assignee.setRole(role);
    users.save(assignee);
    setSlot(team, role, userId);
    teams.save(team);

    audit(actor, companyId, assignAction(role), id,
        Map.of("userId", userId, "role", role.name()), ip);

    return new AssignMemberResult(detail(team), devPassword);
  }

  // --- mapping --------------------------------------------------------------

  private TeamSummaryView summary(Team team) {
    return new TeamSummaryView(
        team.getId(),
        team.getName(),
        slotMember(team.getHrUserId()),
        slotMember(team.getManagerUserId()),
        slotMember(team.getAccountantUserId()),
        isComplete(team),
        users.countByTeamId(team.getId()),
        team.getCreatedAt().toString());
  }

  private TeamDetailView detail(Team team) {
    List<TeamMemberView> members =
        users.findByTeamIdOrderByNameAsc(team.getId()).stream().map(TeamsService::memberView).toList();
    return new TeamDetailView(
        team.getId(),
        team.getName(),
        slotMember(team.getHrUserId()),
        slotMember(team.getManagerUserId()),
        slotMember(team.getAccountantUserId()),
        isComplete(team),
        members.size(),
        team.getCreatedAt().toString(),
        members);
  }

  /** A team is complete when all three slots — HR, Manager, Accountant — are filled (§2). */
  private static boolean isComplete(Team team) {
    return team.getHrUserId() != null
        && team.getManagerUserId() != null
        && team.getAccountantUserId() != null;
  }

  private static String slotUserId(Team team, UserRole role) {
    return switch (role) {
      case HR -> team.getHrUserId();
      case MANAGER -> team.getManagerUserId();
      case ACCOUNTANT -> team.getAccountantUserId();
      default -> null;
    };
  }

  private static void setSlot(Team team, UserRole role, String userId) {
    switch (role) {
      case HR -> team.setHrUserId(userId);
      case MANAGER -> team.setManagerUserId(userId);
      case ACCOUNTANT -> team.setAccountantUserId(userId);
      default -> {
        /* not a team slot */
      }
    }
  }

  /** The other two slot holders (non-null) — a person may hold at most one slot per team. */
  private static List<String> otherSlotUserIds(Team team, UserRole role) {
    return java.util.stream.Stream.of(
            role == UserRole.HR ? null : team.getHrUserId(),
            role == UserRole.MANAGER ? null : team.getManagerUserId(),
            role == UserRole.ACCOUNTANT ? null : team.getAccountantUserId())
        .filter(java.util.Objects::nonNull)
        .toList();
  }

  private static String assignAction(UserRole role) {
    return switch (role) {
      case HR -> "TEAM_HR_ASSIGNED";
      case MANAGER -> "TEAM_MANAGER_ASSIGNED";
      case ACCOUNTANT -> "TEAM_ACCOUNTANT_ASSIGNED";
      default -> "TEAM_MEMBER_ASSIGNED";
    };
  }

  private TeamMemberView slotMember(String userId) {
    if (userId == null) {
      return null;
    }
    return users.findById(userId).map(TeamsService::memberView).orElse(null);
  }

  private static TeamMemberView memberView(User user) {
    return new TeamMemberView(
        user.getId(), user.getName(), user.getEmail(), user.getRole(), user.getStatus());
  }

  // --- helpers --------------------------------------------------------------

  private Team requireTeam(String id, String companyId) {
    return teams.findByIdAndCompanyId(id, companyId).orElseThrow(() -> notFound("Team not found"));
  }

  private UserRole parseTeamRole(String role) {
    if (UserRole.HR.name().equals(role)) {
      return UserRole.HR;
    }
    if (UserRole.MANAGER.name().equals(role)) {
      return UserRole.MANAGER;
    }
    if (UserRole.ACCOUNTANT.name().equals(role)) {
      return UserRole.ACCOUNTANT;
    }
    throw badRequest("role must be HR, MANAGER or ACCOUNTANT");
  }

  private void audit(
      IhrmsPrincipal.User actor,
      String companyId,
      String action,
      String teamId,
      Map<String, Object> metadata,
      String ip) {
    // Partition under the TARGET company (SUPER_ADMIN actor has no company of its own).
    audit.record(new AuditActor("USER", actor.userId(), companyId), action, "Team", teamId, metadata, ip);
  }

  private static boolean isBlank(String s) {
    return s == null || s.isBlank();
  }

  private boolean isProd() {
    return env.acceptsProfiles(Profiles.of("prod"));
  }

  private ResponseStatusException conflict(String message) {
    return new ResponseStatusException(HttpStatus.CONFLICT, message);
  }

  private ResponseStatusException notFound(String message) {
    return new ResponseStatusException(HttpStatus.NOT_FOUND, message);
  }

  private ResponseStatusException badRequest(String message) {
    return new ResponseStatusException(HttpStatus.BAD_REQUEST, message);
  }
}
